Observed counter

tik-stbuehler opened the pull request:
Hi,

I need to read counters from somewhere else, and `Counter` has a hard restriction on only incrementing values - but counters actually are allowed to "reset" to zero, which means the only real requirement for a counter is to have a non-negative value.

`ObservedCounter` is a counter without restrictions, but only provides a `set()` interface to mutate the value.

Also includes a commit for `timestamp_mut()`: I don't want to clone it just to be able to set the timestamp.

It would probably simplify the code if there was an `Arc<(MetricName, Labels)>` wrapper type (not sure what to name it though - `MetricKey`? `MetricIdentifier`?).

tik-stbuehler committed:
timestamp_mut() doesn't need mutable reference

This removes the need to call clone() before timestamp_mut().

tik-stbuehler committed:
add ObservedCounter: a counter for values from external sources

sile commented:
Thank you for your PR.

I think that your motivation is reasonable, but this change introduces a lot of redundant code.
What do you think of adding `set` method to the existing `Count` struct (with `unsafe` declaration)?

> Also includes a commit for `timestamp_mut()`: I don't want to clone it just to be able to set the timestamp.

Sounds good. Thanks.

tik-stbuehler commented:
Hi,

> I think that your motivation is reasonable, but this change introduces a lot of redundant code.
> What do you think of adding `set` method to the existing `Count` struct (with `unsafe` declaration)?

The existing `Counter` has an inner `Value` consisting of an `AtomicF64` and an `AtomicU64`.  For increasing them the race conditions don't matter, but I don't see a way resetting both.

I see two ways solving this:

1. Reduce value to a single `AtomicF64`. Additionally you could make it "wrap" (to zero) at some crazy high value (say `2^52`) to make sure `increment()` still changes the value.
2. Use `Mutex<(f64, u64)>` as value

Would you prefer one of those over adding a separate `ObservedCounter` with redundant code?

sile commented:
Thank you for your explanation.
It seems a more difficult problem than I expected.

BTW, I came up with the following idea to handle race conditions correctly (I'm not sure though):
```diff
diff --git a/src/metrics/counter.rs b/src/metrics/counter.rs
index 3ea2413..fdd984c 100644
--- a/src/metrics/counter.rs
+++ b/src/metrics/counter.rs
@@ -78,6 +78,11 @@ impl Counter {
         self.0.value.increment()
     }
 
+    pub fn set(&self, value: f64) -> Result<()> {
+        self.0.value.set(value);
+        Ok(())
+    }
+
     /// Adds `count` to this counter.
     #[inline]
     pub fn add(&self, count: f64) -> Result<()> {
@@ -277,6 +282,13 @@ impl Value {
     fn add_u64(&self, count: u64) {
         self.u64.add(count);
     }
+
+    fn set(&self, value: f64) {
+        self.f64.update(|old| {
+            let delta = value - (old + self.u64.get() as f64);
+            old + delta.max(0.0)
+        })
+    }
 }
 ```

tik-stbuehler committed:
use std atomics

- requires rustc >= 1.34
- get rid of unsafe f64 -> u64 conversions
- even 32-bit archs usually support 64-bit atomics - let stdlib worry
  about it
- `Relaxed` is good enough: we don't synchronize access to other data
  using the counters, so no ordering with other instructions is
  necessary.

tik-stbuehler commented:
Didn't even realize you had your own atomics, but should have realized there is no standard one for `f64`...

But you probably should use [`std::sync::atomic::AtomicI64`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicI64.html) and [`std::sync::atomic::AtomicU64`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html) (requires `rustc >= 1.34`) - I will create a PR for that (also `Relaxed` will be totally fine as order).

The race condition isn't part of your atomics though: it is because `Counter` uses two atomic values, and can't update both in one "atomic step".

sile committed:
Merge pull request #5 from tik-stbuehler/atomic-fixes

use std atomics

sile commented:
> it is because `Counter` uses two atomic values, and can't update both in one "atomic step".

Yes, you're right. `Counter`  can't update both in one atomic step.
But, in the above diff code (I simplified it today), `Counter` only needs to update `f64` field. 
Thus, in my thought, the consistency of `Counter` isn't broken by the `set` method.

tik-stbuehler commented:
I'm not super excited about using negative `f64` values to counter the `u64` part, but I guess it might work.  But your example code:

1. can't actually decrement in `set()` due to `delta.max(0.0)` - why? I'd probably use `self.f64.add(value.max(0.0) - self.u64.get() as f64)`  (or an `assert!(value >= 0.0)` instead of `.max(0.0)`)

2. I'd read the `u64` value only once.

   Increments to the `u64` and `f64` parts are (and were) never "strictly" ordered, i.e. getting the sum might see an "later" increment to one but not an "earlier" to the other (this is not a real issue with the prometheus semantic of counters), but this means it doesn't matter in concurrent updates (unlikely anyway) whether you refresh the `u64` value in each iteration or not.

   But making the loop smaller increases the chances of it succeeding (given the `Relaxed` order the compiler and the CPU are allowed to optimize it anyway afaics).

3. If the `u64` part wraps around `get()` might return negative values - maybe protect against that with `.max(0.0)`.

I also split off the timestamp mut part, sorry for mixing it in the first place :)

sile committed:
Merge pull request #6 from tik-stbuehler/timestamp-non-mut

timestamp_mut() doesn't need mutable reference

sile commented:
> using negative `f64` values to counter the `u64` part

Hmm. My code doesn't use "negative f64 value".
It simply adds the difference between the value to be set and the current value of `Counter`.
The following code may help your understanding:
```rust
let counter = Counter::new("foo");
counter.add(10); // counter = 10

// Set `value` to `counter`.
// (Except the handling of parallel accesses, the above `set` method does the same work as the following three lines)
let value = 15;
let delta = value - counter.get();  // delta = 5 
counter.add(delta); // counter = 15 = value
```

> (1) can't actually decrement in `set()` due to `delta.max(0.0)`

Because prometheus counters should be monotonically increased, I prohibited negative `delta` values.

> (2) I'd read the u64 value only once.

That makes sense.

> (3) If the `u64` part wraps around `get()` might return negative values

Sorry, I didn't understand this meaning.
The following code doesn't return negative values:
```rust
println!("{}", std::u64::MAX as f64);
// => 18446744073709552000.0

println!("{}", (std::u64::MAX.wrapping_add(1)) as f64);
// => 0.0
```

sile commented:
> counters actually are allowed to "reset" to zero

Sorry, I didn't care this case in the above implementation.

Do you need the ability to reset a counter value?
If not, I think the above `set` method may be sufficient.

tik-stbuehler commented:
As I said in my first post:
> I need to read counters from somewhere else, and Counter has a hard restriction on only incrementing values - but counters actually are allowed to "reset" to zero, which means the only real requirement for a counter is to have a non-negative value.

I don't need the ability to reset it (to zero), I need to ability to set it to **any** non-negative value.

sile commented:
> I don't need the ability to reset it (to zero), I need to ability to set it to any non-negative value.

Prometheus allows counters to be reset to zero. But, IMO, it can be a source of trouble to allow setting any non-negative value to counters. Because, in my understanding, Prometheus handles counter's wrap around as follows:
```
# Pseudocode

counter = 0   # Initialize a counter

counter.set(10000) 
# prometheus reads the counter value and handles the current value as `10,000`

counter.set(9000)  
# prometheus reads the counter value and handles the current value as `19,000` (not `9,000`)
# in some query functions
```

So, I think that it's safer to allow counters to set values only if it doesn't decrease the current values.

tik-stbuehler commented:
PLEASE JUST FUCKING READ MY REQUIREMENTS.

Really.

My program is not the **source** of the counter value.  I don't control it.  I simple need to pass it on.  If the **source** of the counter restarts, it will start from 0 again.

If the source of the value screws it up, then the stats won't look pretty.  You're limited view of what is safe (**not** the way we define the term in rust usually) prevents me using the library in a sane way.

sile commented:
Sorry for my lack of understanding.
Now, I understand your intention. It seems reasonable to create a new struct for this purpose.

sile commented:
I reviewed the code again, and it looks good.

> It would probably simplify the code if there was an `Arc<(MetricName, Labels)>` wrapper type (not sure what to name it though - `MetricKey`? `MetricIdentifier`?).

I like `MetricIdentifier`.
Do you want to apply this simplification within this PR?
(I think that it's okay to separate it into another PR.)

tik-stbuehler commented:
I think it might make this commit smaller to add `MetricIdentifier` first. Might do it in a separate PR.

tik-stbuehler commented:
(While working on `MetricIdentifier`:) The `fn labels_mut(&self)` accessor seems like a bad design to me. It would be a lot easier if those were removed, or if `AtomicImmut` would implement `Clone` (or creation from `Arc<T>`).

tik-stbuehler commented:
Another thing (just realized what the code actually does): I don't think auto-aggregation of metrics is a good thing.  For counters this might lead to decreasing values *without* explizit reset (if one of the aggregated counters is removed)...

My current WIP is in https://github.com/tik-stbuehler/prometrics/tree/metric-identifier and https://github.com/tik-stbuehler/prometrics/tree/observed_counter2 (on top of the other).

sile commented:
> I don't think auto-aggregation of metrics is a good thing.
> For counters this might lead to decreasing values without explizit reset (if one of the aggregated counters is removed)...

Yes. You're right.
Since I use auto-aggregation functionality in several projects, it's difficult to remove it immediately.
However, I would like to reconsider it at the timing the version of `prometrics` is bumped to `v0.2`.

As a workaround, you can disable it by assigning a unique identifier (metric name and label keys) to each metrics.
